enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,236 @@
+
+#### Invalid value
+
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value via the `etcd.operator.openshift.io/v1` API.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendQuotaGiB of type int32
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
+  - With an upper limit of '32GiB'.
+
+```go
+type EtcdSpec struct {
+  // BackendQuotaGiB sets the etcd backend storage size limit in gibibytes. Defaults to 8GiB.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Minimum=8
+	// +kubebuilder:validation:Maximum=32
+	// +kubebuilder:validation:XValidation:rule="self>=oldSelf",message="can't decrease database size"
+	// +openshift:enable:FeatureGate=EtcdBackendQuota
+	// +optional
+  BackendQuotaGiB int32 `json:"backendQuotaGiB"`
+}
+```
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+Since Hypershift does not deploy the cluster etcd operator it can not make use of this feature automatically; it would need to handle detecting the change in the api and rolling out the etcd containers with the new quota. [It configures the quota itself.](https://github.com/openshift/hypershift/blob/main/control-plane-operator/controllers/hostedcontrolplane/etcd/reconcile.go#L408)
+
+#### Standalone Clusters
+
+This change should operate as expected on standalone clusters.
+
+#### Single-node Deployments or MicroShift
+
+This feature should not change the resource consumption of a single-node OpenShift deployment. If an admin decides to increase the quota, the same caveats, from normal OpenShift, about disk space will apply - etcd will not immediately reserve the quota-size, but could grow up to the quota and exhaust disk space in the process.
+
+This enhancement will not directly impact MicroShift as the api will be consume via the Cluster Etcd Operator.
+[Also, changing this value is already supported in the MicroShift configuration](https://github.com/openshift/microshift/blob/main/pkg/config/etcd.go#L18).
+
+### Implementation Details/Notes/Constraints
+
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
+* Remove the parameters from the podspec rendering and replace with the value.
+
+### Risks and Mitigations
+
+* We will need to test larger sizes to ensure that everything works as expected.
+* For a larger database, defragmentation of etcd will take longer which will increase the amount of time that etcd is unavaiable for writes.
+  Etcd will still be available for reads during defragmentation, so there should be little impact on the apiserver's availability.
+* Downgrades may be impacted if:
+  1. The downgrade is to a version without this feature.
```

Comment:
Probably worth considering a way to resolve this problem in the future, not a blocker atm.

Reply:
One potential approach would be to add a check that this is not set in a version prior to when this feature goes GA. To prevent that kind of downgrade, or even block it.